File: C++/Sorting(Non-Comparison)/BucketSort.cpp
#include <bits/stdc++.h>
using namespace std;

void bucketSort(vector<float>& arr)
{
    if (arr.empty()) return;
    int n = arr.size();
    float minVal = *min_element(arr.begin(), arr.end());
    float maxVal = *max_element(arr.begin(), arr.end());
    vector<vector<float>> buckets(n);
    for (int i = 0; i < n; i++)
    {
        int idx = (int)((arr[i] - minVal) / (maxVal - minVal + 1) * n);
        buckets[idx].push_back(arr[i]);
    }
    arr.clear();
    for (int i = 0; i < n; i++)
    {
        sort(buckets[i].begin(), buckets[i].end());
        arr.insert(arr.end(), buckets[i].begin(), buckets[i].end());
    }
}

int main()
{
    vector<float> arr = {0.42, 4.21, 1.23, 0.12, 2.34, 3.14, 2.71};
    cout << "Original array: ";
    for (float n : arr)
        cout << n << " ";
    cout << endl;
    bucketSort(arr);
    cout << "Sorted array: ";
    for (float n : arr)
        cout << n << " ";
    cout << endl;
    return 0;
}

File: Python/Sorting(Non-Comparison)/BucketSort.py
def bucket_sort(arr):
    if not arr:
        return arr
    min_val = min(arr)
    max_val = max(arr)
    bucket_count = len(arr)
    buckets = [[] for _ in range(bucket_count)]
    for num in arr:
        index = int((num - min_val) / (max_val - min_val + 1) * bucket_count)
        buckets[index].append(num)
    arr.clear()
    for bucket in buckets:
        arr.extend(sorted(bucket))
    return arr

if __name__ == "__main__":
    arr = [0.42, 4.21, 1.23, 0.12, 2.34, 3.14, 2.71]
    print("Original array:", arr)
    bucket_sort(arr)
    print("Sorted array:", arr)
